src/libstd/path.rs

```diff
@@ -0,0 +1,2415 @@
+            (Some(Component::Empty), Some(Component::Prefix(_))) => None,
+            (Some(Component::Prefix(_)), None) => None,
+            (Some(Component::RootDir), Some(Component::Prefix(_))) => None,
+            _ => rest
+        }
+    }
+
+    /// The final component of the path, if it is a normal file.
+    ///
+    /// If the path terminates in `.`, `..`, or consists solely or a root of
+    /// prefix, `file` will return `None`.
+    pub fn file_name(&self) -> Option<&OsStr> {
+        self.components().next_back().and_then(|p| match p {
+            Component::Normal(p) => Some(p.as_os_str()),
+            _ => None
+        })
+    }
+
+    /// Returns a path that, when joined onto `base`, yields `self`.
+    pub fn relative_from<'a, P: ?Sized>(&'a self, base: &'a P) -> Option<&Path> where
+        P: AsPath
+    {
+        iter_after(self.components(), base.as_path().components()).map(|c| c.as_path())
+    }
+
+    /// Determines whether `base` is a prefix of `self`.
+    pub fn starts_with<P: ?Sized>(&self, base: &P) -> bool where P: AsPath {
+        iter_after(self.components(), base.as_path().components()).is_some()
+    }
+
+    /// Determines whether `base` is a suffix of `self`.
+    pub fn ends_with<P: ?Sized>(&self, child: &P) -> bool where P: AsPath {
+        iter_after(self.components().rev(), child.as_path().components().rev()).is_some()
+    }
+
+    /// Extract the stem (non-extension) portion of `self.file()`.
+    ///
+    /// The stem is:
+    ///
+    /// * None, if there is no file name;
+    /// * The entire file name if there is no embedded `.`;
+    /// * The entire file name if the file name begins with `.` and has no other `.`s within;
+    /// * Otherwise, the portion of the file name before the final `.`
+    pub fn file_stem(&self) -> Option<&OsStr> {
+        self.file_name().map(split_file_at_dot).and_then(|(before, after)| before.or(after))
+    }
+
+    /// Extract the extension of `self.file()`, if possible.
+    ///
+    /// The extension is:
+    ///
+    /// * None, if there is no file name;
+    /// * None, if there is no embedded `.`;
+    /// * None, if the file name begins with `.` and has no other `.`s within;
+    /// * Otherwise, the portion of the file name after the final `.`
+    pub fn extension(&self) -> Option<&OsStr> {
+        self.file_name().map(split_file_at_dot).and_then(|(before, after)| before.and(after))
+    }
+
+    /// Creates an owned `PathBuf` with `path` adjoined to `self`.
+    ///
+    /// See `PathBuf::push` for more details on what it means to adjoin a path.
+    pub fn join<P: ?Sized>(&self, path: &P) -> PathBuf where P: AsPath {
+        let mut buf = self.to_path_buf();
+        buf.push(path);
+        buf
+    }
+
+    /// Creates an owned `PathBuf` like `self` but with the given file name.
+    ///
+    /// See `PathBuf::set_file_name` for more details.
+    pub fn with_file_name<S: ?Sized>(&self, file_name: &S) -> PathBuf where S: AsOsStr {
```

Comment:
I find it interesting that this takes `AsOsStr` while other methods take `AsPath`. It is consistent with filename-related methods taking `OsStr`-like things, but it is somewhat odd in terms of `AsPath` and `AsOsStr` basically being the same trait.